grammar.js

```diff
@@ -1091,15 +1100,17 @@ function sep1 (rule, separator) {
   return seq(rule, repeat(seq(separator, rule)))
 }
 
+// This can be merged with whiteSpacedKeyword, keeping for now.
 function blockStructureEnding ($, structType) {
   const obj = prec.right(seq(
-    caseInsensitive('end'),
-    optional(seq(
-      caseInsensitive(structType),
-      optional($.identifier)
-    )),
-    $._end_of_statement
-  ))
+    alias(choice(
+      seq(
+        caseInsensitive('end',aliasAsWord=false),
+        optional(caseInsensitive(structType,aliasAsWord=false))),
+      caseInsensitive('end'+structType,aliasAsWord=false)),
+      'end'+structType),
+    optional($.identifier),
+    optional($._end_of_statement)))
```

Comment:
@oponkork does this EOS token need to be optional for any particular reason? It doesn't break the grammar, but it does allow invalid syntax to be parsed without erroring out. Tree-sitter isn't meant to validate code necessarily but it's preferable for invalid syntax to throw a parse error.

Reply:
@stadelmanma I made it optional by mistake, while changing the code. Fixed.